### DISCUSSION TITLE
How do I build Bun for production

### DISCUSSION THREAD

**Discussion by NoelJacob** [oven-sh/bun#2354] [Q&A]
I want to compile Bun from scratch for production without Docker or downloading any previous Bun binary. How to?

It might be okay to download a compiled JSC from the repo but also can you tell how to compile that repo too.

**Comment by vjpr** [comment]
https://bun.sh/docs/project/developing

**Reply by NoelJacob** [reply]
I've read it a few times and it doesn't show building for production.

**Comment by hasanyasin** [comment]
@NoelJacob,

I'm not sure if I understand your question correctly. I assume that you want to build the `bun` binary from scratch. You'll need `zig` to compile, then JavaScriptCore and all its dependencies ready on the system.

If you want to follow all the steps yourself, just check how the dev container is built:

https://github.com/oven-sh/bun/blob/main/Dockerfile.devcontainer

There, you'll see all the variables set, all the commands ran to have the binary compiled successfully.

**Reply by NoelJacob** [reply]
The thing I want to do is this - I have zig, and I want to compile JavaScriptCore for production and then bun for production. How do I do this without a docker?

I am asking this to (try to) create a config/installation file for a few package managers that installs bun.

**Comment by hasanyasin** [comment]
The linked Dockerfile is what runs to build the container with all the steps. That is your recipe. As you see at the from (`FROM ... ubuntu:22.04`), it is an Ubuntu system so you can run the exact commands on an Ubuntu, and probably any other Debian-based machine.

A Dockerfile is usually quite easy to convert to a bash script. It's all commands and some env/variable assignments.

**Reply by NoelJacob** [reply]
I'm looking at the Dockerfiles but is there anyway I can build with running `make` commands after setting variables. I also want to compile JSC myself because it seems possible with this Makefile. I think now the preferred way to get into development is through `make setup`.

**Comment by jelly8173** [comment]
So its not actually possible to build Bun if you does not possess bun binary already, because it is used in build process for code generation:
```
$ ninja -v
...
[2/382] cd /home/jelly8173/temp/bun && bun run src/codegen/generate-jssink.ts /home/jelly8173/temp/bun/build/codegen
```

And as you can see, developers does not try to make build scripts compatible with anything except Bun itself, as they use internal bun modules. Right now only for spawning processes and file IO, but that situation will get worse if they don't change this direction.

**Comment by COD3HUNT3R** [comment]
Maybe this can help. @LowByteFox worked on native ports for FreeBSD

https://github.com/oven-sh/bun/pull/7218
https://gist.github.com/LowByteFox/a4e86c64aa3eaae6f79efb293faefe74
